frame/vesting/src/lib.rs

```diff
@@ -312,105 +369,382 @@ pub mod pallet {
 	fn vesting_balance(who: &T::AccountId) -> Option<BalanceOf<T>> {
 		if let Some(v) = Self::vesting(who) {
 			let now = <frame_system::Pallet<T>>::block_number();
-			let locked_now = v.locked_at::<T::BlockNumberToBalance>(now);
-			Some(T::Currency::free_balance(who).min(locked_now))
+			let total_locked_now = v.iter().fold(Zero::zero(), |total, schedule| {
+				schedule.locked_at::<T::BlockNumberToBalance>(now).saturating_add(total)
+			});
+			Some(T::Currency::free_balance(who).min(total_locked_now))
 		} else {
 			None
 		}
 	}
 
 	/// Adds a vesting schedule to a given account.
 	///
-	/// If there already exists a vesting schedule for the given account, an `Err` is returned
-	/// and nothing is updated.
+	/// If the account has `MaxVestingSchedules`, an Error is returned and nothing
+	/// is updated.
 	///
 	/// On success, a linearly reducing amount of funds will be locked. In order to realise any
 	/// reduction of the lock over time as it diminishes, the account owner must use `vest` or
 	/// `vest_other`.
 	///
 	/// Is a no-op if the amount to be vested is zero.
+	/// NOTE: does not validate schedule params.
 	fn add_vesting_schedule(
 		who: &T::AccountId,
 		locked: BalanceOf<T>,
 		per_block: BalanceOf<T>,
-		starting_block: T::BlockNumber
+		starting_block: T::BlockNumber,
 	) -> DispatchResult {
-		if locked.is_zero() { return Ok(()) }
-		if Vesting::<T>::contains_key(who) {
-			Err(Error::<T>::ExistingVestingSchedule)?
-		}
-		let vesting_schedule = VestingInfo {
-			locked,
-			per_block,
-			starting_block
-		};
-		Vesting::<T>::insert(who, vesting_schedule);
-		// it can't fail, but even if somehow it did, we don't really care.
-		let res = Self::update_lock(who.clone());
-		debug_assert!(res.is_ok());
+		if locked.is_zero() { return Ok(()); }
+
+		let vesting_schedule = VestingInfo::new::<T>(locked, per_block, starting_block);
+		let mut schedules = Self::vesting(who).unwrap_or_default();
+
+		// NOTE: we must push the new schedule so that `report_schedule_updates`
+		// will give the correct new locked amount.
+		ensure!(schedules.try_push(vesting_schedule).is_ok(), Error::<T>::AtMaxVestingSchedules);
+
+		let (schedules, locked_now) =
+			Self::report_schedule_updates(schedules.to_vec(), VestingAction::Passive);
+
+		debug_assert!(
+			locked_now > Zero::zero() && schedules.len() > 0 ||
+				locked_now == Zero::zero() && schedules.len() == 0
+		);
+		Self::write_vesting(&who, schedules)?;
+		Self::write_lock(who, locked_now);
+
+		Ok(())
+	}
+
+	// Ensure we can call `add_vesting_schedule` without error. This should always
+	// be called prior to `add_vesting_schedule`.
+	//
+	// NOTE: expects schedule param validation has been done due to different
+	// scenarios having varying requirements.
+	fn can_add_vesting_schedule(
+		who: &T::AccountId,
+		_locked: BalanceOf<T>,
+		_per_block: BalanceOf<T>,
+		_starting_block: T::BlockNumber,
+	) -> DispatchResult {
```

Comment:
should fail if per_block is zero

Reply:
24214ac2e0f049504c29057a88856c9d7ca02da5